### ISSUE TITLE
revert #7957 to validate the performance regression of q7 and try to improve if true

### ISSUE THREAD

**Issue by lmatz** [risingwavelabs/risingwave#8510]
![SCR-20230313-paf](https://user-images.githubusercontent.com/5791930/224671974-f22cc11d-451d-43a5-aeed-aa1d5cee7108.png)

Meanwhile q7-rewrite which uses `group top-n` does not change performance much. Didn't see other queries also affected.

**Comment by lmatz** [comment]
check the performance of the queries that involve join:
https://buildkite.com/risingwave-test/nexmark-benchmark/builds/571#0186de49-a190-42f2-a56f-4b31e4130513 

Let's wait and see more results...

**Comment by lmatz** [comment]
Probably it still depends on the real workload to determine whether it improves or makes the performance worse?
Then how about making it a per-query option? 
Query hints, or session variables (hard to control if there are multiple joins in a single query)

**Comment by fuyufjh** [comment]
Interesting 🤔 cc. @hzxa21 Any ideas about the reason?

**Comment by hzxa21** [comment]
TL;DR, I think in the perf benchmark, the locality between the two stream is worse than expected so refilling cache on write in the join executor cause eviction on the old entries that are still needed, leading to higher operator cache miss. cc @KeXiangWang This is different from what we have seen before. Any thoughts?

Findings after looking at grafana (up: [02/23](https://g-2927a1b4d9.grafana-workspace.us-east-1.amazonaws.com/d/EpkBw5W4k/risingwave-benchmark?orgId=1&from=1677173722000&to=1677175376000&to=&var-namespace=nexmark-0-3-4-5-7-daily-20230223) no `may_exist` and join cache refill on write, down: [02/24](https://g-2927a1b4d9.grafana-workspace.us-east-1.amazonaws.com/d/EpkBw5W4k/risingwave-benchmark?orgId=1&from=1677260894000&to=1677262729000&to=&var-namespace=nexmark-0-3-4-5-7-daily-20230224)):
1. Aggregation Cached Keys is way smaller in 02/24
![image](https://user-images.githubusercontent.com/5518566/225291878-9c8dd24c-1610-4d20-b74b-17a736baf7c9.png)
![image](https://user-images.githubusercontent.com/5518566/225291931-335abc08-0c1d-43dc-baca-1743ef96d60a.png)

2. Join Cached Entries is also way smaller in 02/24
![image](https://user-images.githubusercontent.com/5518566/225292704-3f9a4bd5-5c07-4840-881a-93408c42ac84.png)
![image](https://user-images.githubusercontent.com/5518566/225292732-e3d056bb-d977-4db4-baae-a80b7a1258a9.png)

3. Join cache miss rate is higher. Cache misses when insert entries in the left table is high and the miss keys are most likely existing keys (`may_exists` true rate is high) in 02/24
![image](https://user-images.githubusercontent.com/5518566/225294547-8f622753-f00d-4953-8786-bbc418021248.png)
![image](https://user-images.githubusercontent.com/5518566/225294754-488cc311-e304-4941-b831-d6a01a23a383.png)

4. Read duration of the join left and right table are way higher in 02/24 
![image](https://user-images.githubusercontent.com/5518566/225296918-d0e0b416-6280-40f9-ab26-bcf838ae8db0.png)
![image](https://user-images.githubusercontent.com/5518566/225296952-d9be7736-f793-4e70-ac46-c87b0cd91494.png)

**Comment by KeXiangWang** [comment]
I agree with Patrick. Nexmark Q7 is join `price` of bids and the `maxprice` of bids within a interval. With refilling, all the bid will be cache in the join cache. But you can image, `maxprice` stream's event is sparse and only read several keys of `price`. So, I believe the throughput decrease because of two reason:
1. Unnecessary call of `may_exist` when inserting a `price` event intruduce overhead.
2. `price` event occupy too much cache, causing cache miss and thus latences.

Like @lmatz mentioned, we may probably make decision depends on task and data patterns😕.

**Comment by lmatz** [comment]
Later, we found that the 8 kafka-partition nexmark is generated by 16 threads, which can mess up the locality.
Now I suppose this factor may affect the effectiveness of refilling.

I will try to re-enable it later and make it a configurable choice.
